File: samples/tutorials-xamarin/tutorials-3d/tutorial-04/MainActivity.cs
using System;
using System.Timers;
using Android.App;
using Android.Graphics;
using Android.OS;
using Android.Views;
using AndroidX.AppCompat.App;
using SciChart.Charting.Visuals;
using SciChart.Charting3D.Model.DataSeries.Xyz;
using SciChart.Charting3D.Modifiers;
using SciChart.Charting3D.Visuals;
using SciChart.Charting3D.Visuals.Axes;
using SciChart.Charting3D.Visuals.PointMarkers;
using SciChart.Charting3D.Visuals.RenderableSeries.Scatter;
using SciChart.Core.Model;

namespace Tutorial_3d_04
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme.NoActionBar", MainLauncher = true)]
    public class MainActivity : AppCompatActivity
    {
        private readonly object _syncRoot = new object();

        // <DataSeriesSetup>
        private const int PointsCount = 200;
        private const long TimerInterval = 10;

        private Timer _timer;

        private SciChartSurface3D surface;

        private readonly DoubleValues xValues = new DoubleValues();
        private readonly DoubleValues yValues = new DoubleValues();
        private readonly DoubleValues zValues = new DoubleValues();
    
        private readonly XyzDataSeries3D<double, double, double> dataSeries = new XyzDataSeries3D<double, double, double>();
        // </DataSeriesSetup>

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Xamarin.Essentials.Platform.Init(this, savedInstanceState);

            SciChartSurface.SetRuntimeLicenseKey("");
            SetContentView(Resource.Layout.activity_main);

            surface = new SciChartSurface3D(this);
            var chartLayout = FindViewById<ViewGroup>(Resource.Id.chart_layout);
            chartLayout.AddView(surface);

            // <AppendDataSeries>
            for (int i = 0; i < PointsCount; i++)
            {
                xValues.Add(GetGaussianRandomNumber(5.0, 1.5));
                yValues.Add(GetGaussianRandomNumber(5.0, 1.5));
                zValues.Add(GetGaussianRandomNumber(5.0, 1.5));
            }
            dataSeries.Append(xValues, yValues, zValues);
            // </AppendDataSeries>

            var rSeries = new ScatterRenderableSeries3D()
            {
                DataSeries = dataSeries,
                PointMarker = new SpherePointMarker3D()
                {
                    Fill = new Color(0x32, 0xCD, 0x32, 0xFF),
                    Size = 10
                }
            };

            using (surface.SuspendUpdates())
            {
                surface.XAxis = new NumericAxis3D();
                surface.YAxis = new NumericAxis3D();
                surface.ZAxis = new NumericAxis3D();

                surface.RenderableSeries.Add(rSeries);

                // <AddModifiers>
                surface.ChartModifiers.Add(new OrbitModifier3D { ExecuteOnPointerCount = 2 });
                surface.ChartModifiers.Add(new ZoomExtentsModifier3D());
                surface.ChartModifiers.Add(new PinchZoomModifier3D());
                surface.ChartModifiers.Add(new TooltipModifier3D { CrosshairMode = CrosshairMode.Lines });
                // </AddModifiers>
            }

            // <CreateTimer>
            _timer = new Timer(TimerInterval) { AutoReset = true };
            _timer.Elapsed += OnTick;
            _timer.Start();
            // </CreateTimer>
        }   

        // <UpdateData>
        private Random _random = new Random();

        private void OnTick(object sender, ElapsedEventArgs e)
        {
            lock (_syncRoot)
            {
                for (int i = 0; i < PointsCount; i++)
                {
                    var xValue = xValues.Get(i) + _random.NextDouble() - 0.5;
                    var yValue = yValues.Get(i) + _random.NextDouble() - 0.5;
                    var zValue = zValues.Get(i) + _random.NextDouble() - 0.5;

                    xValues.Set(i, xValue);
                    yValues.Set(i, yValue);
                    zValues.Set(i, zValue);
                }

                using (surface.SuspendUpdates())
                {
                    dataSeries.UpdateRangeXyzAt(0, xValues, yValues, zValues);
                }
            }
        }
        // </UpdateData>

        private double GetGaussianRandomNumber(double mean, double stdDev)
        {
            var random = new Random();

            double u1 = random.NextDouble();
            double u2 = random.NextDouble();

            var randStdNormal = Math.Sqrt(-2.0 * Math.Log10(u1)) * Math.Sin(2.0 * Math.PI * u2);
            return mean * stdDev * randStdNormal;
        }
    }
}

File: samples/first-app/3D/src/main/java/com/example/firstscichartapp/MainActivity.kt
package com.example.firstscichartapp

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.ui.AppBarConfiguration
import com.example.firstscichartapp.databinding.ActivityMainBinding
import com.scichart.charting.model.dataSeries.XyDataSeries
import com.scichart.charting.visuals.SciChartSurface
import com.scichart.charting.visuals.axes.IAxis
import com.scichart.charting.visuals.axes.NumericAxis
import com.scichart.charting.visuals.renderableSeries.FastLineRenderableSeries
import com.scichart.charting3d.model.dataSeries.xyz.XyzDataSeries3D
import com.scichart.charting3d.visuals.SciChartSurface3D
import com.scichart.charting3d.visuals.axes.NumericAxis3D
import com.scichart.charting3d.visuals.renderableSeries.pointLine.PointLineRenderableSeries3D
import com.scichart.core.framework.UpdateSuspender
import com.scichart.core.model.DoubleValues
import kotlin.math.cos
import kotlin.math.sin

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    init {
        SciChartSurface.setRuntimeLicenseKey("")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val surface = findViewById<SciChartSurface3D>(R.id.surface)

        // <AddingAxesToTheSciChartSurface3D>
        UpdateSuspender.using(surface) {
            surface.xAxis = NumericAxis3D()
            surface.yAxis = NumericAxis3D()
            surface.zAxis = NumericAxis3D()
        }
        // </AddingAxesToTheSciChartSurface3D>

        //<Adding3DRenderableSeries>
        val dataSeries = XyzDataSeries3D(Double::class.javaObjectType, Double::class.javaObjectType, Double::class.javaObjectType)

        for (i in 0 until 100) {
            val x = 5 * sin(i.toDouble())
            val y = i.toDouble()
            val z = 5 * cos(i.toDouble())
            dataSeries.append(x, y, z)
        }

        val rSeries = PointLineRenderableSeries3D().apply {
            this.dataSeries = dataSeries
            strokeThickness = 3.0f
        }

        surface.renderableSeries.add(rSeries)
        //</Adding3DRenderableSeries>
    }
}

/*
// <FinalExampleCode>
class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    init {
        SciChartSurface.setRuntimeLicenseKey("")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val surface = findViewById<SciChartSurface3D>(R.id.surface)

        // <AddingAxesToTheSciChartSurface3D>
        UpdateSuspender.using(surface) {
            surface.xAxis = NumericAxis3D()
            surface.yAxis = NumericAxis3D()
            surface.zAxis = NumericAxis3D()
        }
        // </AddingAxesToTheSciChartSurface3D>

        //<Adding3DRenderableSeries>
        val dataSeries = XyzDataSeries3D(Double::class.javaObjectType, Double::class.javaObjectType, Double::class.javaObjectType)

        for (i in 0 until 100) {
            val x = 5 * sin(i.toDouble())
            val y = i.toDouble()
            val z = 5 * cos(i.toDouble())
            dataSeries.append(x, y, z)
        }

        val rSeries = PointLineRenderableSeries3D().apply {
            this.dataSeries = dataSeries
            strokeThickness = 3.0f
        }

        surface.renderableSeries.add(rSeries)
        //</Adding3DRenderableSeries>
    }
}
// </FinalExampleCode>
 */